bpf/bpf_host.c

```diff
@@ -538,12 +554,30 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 	/* Lookup IPv4 address in list of local endpoints and host IPs */
 	ep = lookup_ip4_endpoint(ip4);
 	if (ep) {
+		bool l2_hdr_required __maybe_unused = true;
+		int l3_off __maybe_unused = ETH_HLEN;
+
 		/* Let through packets to the node-ip so they are processed by
 		 * the local ip stack.
 		 */
 		if (ep->flags & ENDPOINT_F_HOST)
 			return CTX_ACT_OK;
 
+#ifdef ENABLE_HOST_ROUTING
+		/* add L2 header for L2-less interface, such as cilium_wg0 */
+		ret = maybe_add_l2_hdr(ctx, ep->ifindex, &l2_hdr_required);
+		if (ret != 0)
+			return ret;
+		if (l2_hdr_required && ETH_HLEN == 0) {
+			/* l2 header is added */
+			l3_off += __ETH_HLEN;
+			if (!____revalidate_data_pull(ctx, &data, &data_end,
+						      (void **)&ip4, sizeof(*ip4),
+						      false, l3_off))
+				return DROP_INVALID;
+		}
+#endif
```

Comment:
Could this check be done unconditionally instead of via ifdefs?

Reply:
I think it could be done, since there's no side effect for normal L3-L3 redirect, or even host routing is off, this `maybe_add_l2_hdr` still does no harm. Julian, how do you like it? @julianwiedmann

Reply:
I don't see the advantage - if we only require it for host-routing, why burden the BPF program on eg. older kernels with the additional complexity?